File: android/src/main/kotlin/co/verifyspeed/flutter_verifyspeed_plugin/FlutterVerifyspeedPlugin.kt
package co.verifyspeed.flutter_verifyspeed_plugin

import android.app.Activity
import co.verifyspeed.android.VerifySpeed
import co.verifyspeed.android.VerifySpeedError
import co.verifyspeed.android.VerifySpeedErrorType
import co.verifyspeed.android.VerifySpeedListener
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.future.await
import kotlinx.coroutines.launch
import java.lang.ref.WeakReference

/** FlutterVerifySpeedPlugin */
class FlutterVerifyspeedPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private val activity get() = activityReference.get()
    private var activityReference = WeakReference<Activity>(null)

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "verifyspeed_channel")

        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val arguments = call.arguments as HashMap<*, *>?
        if (activity == null) {
            result.success(
                mapOf(
                    "error" to "Activity not found",
                    "errorType" to VerifySpeedErrorType.ActivityNotSet
                )
            )

            return
        }

        when (call.method) {
            "initialize" -> {
                handleException({
                    val clientKey = arguments!!["clientKey"] as String

                    VerifySpeed.setClientKey(clientKey)

                    val methods = VerifySpeed.initialize().await()

                    val methodsJson = """
              {
                "availableMethods": ${
                        methods.map { method ->
                            """{"methodName":"${method.methodName}","displayName":"${method.displayName}"}"""
                        }
                    }
              }
            """.trimIndent()

                    result.success(methodsJson)
                }, result)
            }


            "verifyPhoneNumberWithDeepLink" -> {
                handleException(
                    {
                        val deepLink = arguments!!["deepLink"] as String
                        val verificationKey = arguments["verificationKey"] as String
                        val redirectToStore = arguments["redirectToStore"] as? Boolean ?: true

                        VerifySpeed.setActivity(activity!!)

                        VerifySpeed.verifyPhoneNumberWithDeepLink(
                            deepLink,
                            verificationKey,
                            redirectToStore,
                            getVerificationListener(result),
                        )
                    },
                    result
                )
            }

            "verifyPhoneNumberWithOtp" -> {
                handleException({
                    val verificationKey = arguments!!["verificationKey"] as String
                    val phoneNumber = arguments["phoneNumber"] as String

                    VerifySpeed.verifyPhoneNumberWithOtp(
                        phoneNumber,
                        verificationKey,
                    ).await()

                    result.success(null)
                }, result)
            }

            "validateOtp" -> {
                handleException(
                    {
                        val verificationKey: String = arguments!!["verificationKey"] as String
                        val otpCode: String = arguments["otpCode"] as String

                        VerifySpeed.setActivity(activity!!)

                        VerifySpeed.validateOTP(
                            otpCode,
                            verificationKey,
                            getVerificationListener(result),
                        ).await()
                    },
                    result,
                )
            }

            "notifyOnResumed" -> {
                handleException(
                    { VerifySpeed.notifyOnResumed() },
                    result
                )
            }

            "checkInterruptedSession" -> {
                handleException({
                    VerifySpeed.setActivity(activity!!)

                    VerifySpeed.checkInterruptedSession { token ->
                        result.success(mapOf("token" to token))
                    }.await()
                }, result)
            }
        }
    }

    private fun getVerificationListener(result: MethodChannel.Result): VerifySpeedListener {
        return object : VerifySpeedListener {
            override fun onSuccess(token: String) {
                result.success(mapOf("token" to token))
            }

            override fun onFail(error: VerifySpeedError) {
                result.success(mapOf("error" to error.message, "errorType" to error.type.name))
            }
        }
    }

    private fun handleException(func: suspend () -> Unit, result: MethodChannel.Result) {
        CoroutineScope(Dispatchers.Main).launch {
            try {
                func()
            } catch (e: VerifySpeedError) {
                result.success(mapOf("error" to e.message, "errorType" to e.type.name))
            } catch (e: Exception) {
                result.success(mapOf("error" to e.message, "errorType" to "Unknown"))
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityReference = WeakReference(binding.activity)
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivity() {
        activityReference.clear()
        channel.setMethodCallHandler(null)
    }
}

File: ios/Classes/FlutterVerifyspeedPlugin.swift
import Flutter
import UIKit
import VerifySpeed_IOS

public class FlutterVerifyspeedPlugin: NSObject, FlutterPlugin {
    
    private var channel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var navigationController: UINavigationController!
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = FlutterVerifyspeedPlugin()
        
        instance.channel = FlutterMethodChannel(name: "verifyspeed_channel", binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: instance.channel!)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? [String: Any]
        
        switch call.method {

        case "initialize":
            let clientKey = arguments?["clientKey"] as! String
            
            VerifySpeed.shared.setClientKey(clientKey)
            do {
                try VerifySpeed.shared.initialize() { methods in
                    let methodsArray = methods.map { method in
                        """
                        {"methodName":"\(method.methodName)",
                        "displayName":"\(method.displayName)"}
                        """
                    }.joined(separator: ",")
                    
                    let methodsJson = """
                    {
                        "availableMethods": [\(methodsArray)]
                    }
                    """
                    
                    result(methodsJson)
                }
            } catch let error as VerifySpeedError {
                result(["error" : error.message, "errorType" : error.type.name])
            } catch {
                result(["error" : error.localizedDescription, "errorType" : VerifySpeedErrorType.unknown.name])
            }

        case "verifyPhoneNumberWithDeepLink":
            let deeplink = arguments?["deepLink"] as! String
            let verificationKey = arguments?["verificationKey"] as! String
            let redirectToStore = arguments?["redirectToStore"] as? Bool ?? true
            
            let listener = VerifySpeedListenerHandler(result: result)
            
            VerifySpeed.shared.verifyPhoneNumberWithDeepLink(
                deeplink: deeplink,
                verificationKey: verificationKey,
                redirectToStore: redirectToStore,
                callBackListener: listener
            )
            
        case "verifyPhoneNumberWithOtp":
            let phoneNumber = arguments?["phoneNumber"] as! String
            let verificationKey = arguments?["verificationKey"] as! String
            
            VerifySpeed.shared.verifyPhoneNumberWithOtp(
                phoneNumber: phoneNumber,
                verificationKey: verificationKey
            ) { response in
                switch response {
                case .success(let otpResponse):
                    result([
                        "verificationKey": otpResponse.verificationKey,
                        "sentByMethodName": otpResponse.sentByMethodName,
                        "sentByMethodDisplay": otpResponse.sentByMethodDisplay,
                        "nextSendAvailableAt": otpResponse.nextSendAvailableAt
                    ])
                    break
                    
                case .failure(let error):
                    result(["error" : error.message, "errorType" : error.type.name])
                    break
                }
            }
            
        case "sendNextDynamicOtp":
            let verificationKey = arguments?["verificationKey"] as! String
            
            VerifySpeed.shared.sendNextDynamicOtp(
                verificationKey: verificationKey
            ) { response in
                switch response {
                case .success(let otpResponse):
                    result([
                        "verificationKey": otpResponse.verificationKey,
                        "sentByMethodName": otpResponse.sentByMethodName,
                        "sentByMethodDisplay": otpResponse.sentByMethodDisplay,
                        "nextSendAvailableAt": otpResponse.nextSendAvailableAt
                    ])
                    break
                    
                case .failure(let error):
                    result(["error" : error.message, "errorType" : error.type.name])
                    break
                }
            }
            
        case "validateOtp":
            let otpCode = arguments?["otpCode"] as! String
            let verificationKey = arguments?["verificationKey"] as! String
            
            let listener = VerifySpeedListenerHandler(result: result)
            
            VerifySpeed.shared.validateOTP(
                otpCode: otpCode,
                verificationKey: verificationKey,
                callBackListener: listener
            )
            
        case "notifyOnResumed":
            VerifySpeed.shared.notifyOnResumed()
            
        case "checkInterruptedSession":
            VerifySpeed.shared.checkInterruptedSession() { token in
                if let token = token {
                    result(["token": token])
                } else {
                    result(nil)
                }
            }
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

class VerifySpeedListenerHandler : VerifySpeedListener {
    init(result: @escaping FlutterResult) {
        self.result = result
    }
    
    let result: FlutterResult
    
    func onFail(error: VerifySpeedError) {
        result(["error" : error.message, "errorType" : error.type.name])
    }
    
    func onSuccess(token: String) {
        result(["token" : token])
    }
}
